src/raftstore/coprocessor/split_check/half.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rocksdb::DB;
+
+use util::config::ReadableSize;
+
+use super::super::{Coprocessor, ObserverContext, SplitCheckObserver};
+use super::Status;
+
+const BUCKET_NUMBER_LIMIT: usize = 1024;
+const BUCKET_SIZE_LIMIT_MB: u64 = 512;
+
+#[derive(Default)]
+pub struct HalfStatus {
+    buckets: Vec<Vec<u8>>,
+    cur_bucket_size: u64,
+    each_bucket_size: u64,
+}
+
+impl HalfStatus {
+    fn new(each_bucket_size: u64) -> HalfStatus {
+        HalfStatus {
+            each_bucket_size: each_bucket_size,
+            ..Default::default()
+        }
+    }
+
+    pub fn push_data(&mut self, key: &[u8], value_size: u64) {
+        let current_len = key.len() as u64 + value_size;
+        if self.buckets.is_empty() || self.cur_bucket_size >= self.each_bucket_size {
+            self.buckets.push(key.to_vec());
+            self.cur_bucket_size = 0;
+        }
+        self.cur_bucket_size += current_len;
+    }
+
+    pub fn split_key(self) -> Option<Vec<u8>> {
+        let mid = self.buckets.len() / 2;
+        if mid == 0 {
+            None
+        } else {
+            self.buckets.get(mid).cloned()
+        }
+    }
+}
+
+pub struct HalfCheckObserver {
+    half_split_bucket_size: u64,
+}
+
+impl HalfCheckObserver {
+    pub fn new(region_size_limit: u64) -> HalfCheckObserver {
+        let mut half_split_bucket_size = region_size_limit / BUCKET_NUMBER_LIMIT as u64;
```

Comment:
Maybe we need to check the `region_size_limit`, make sure it is larger than the `region_size_limit`.